File: office_convert/orchestrator.py
"""Per-request orchestration: probe → plan → dispatch → merge → stream.

Implements FR-3 through FR-7, FR-9, FR-10. The end-to-end pipeline from
business-logic-model.md §1. Yields PDF bytes as an async generator.
"""

from __future__ import annotations

import asyncio
import hashlib
import logging
import time
from collections.abc import AsyncIterator
from contextlib import suppress
from pathlib import Path
from typing import TYPE_CHECKING

from office_convert import aspose_worker, qpdf
from office_convert.cache import CacheManager
from office_convert.chunk_planner import adaptive_max_pages, chunk_sha256, plan_chunks, subdivide
from office_convert.errors import OOMError, SubdivisionFloorError
from office_convert.logging import emit_event
from office_convert.probe import probe as do_probe
from office_convert.worker_pool import WorkerPool, pool_mode_available
from office_convert.types import (
    Chunk,
    ConversionOptions,
    ConversionResult,
    FormatName,
    ProbeResult,
)

if TYPE_CHECKING:
    from office_convert.config import Settings

log = logging.getLogger(__name__)


async def convert_job(
    request_id: str,
    input_path: Path,
    format: FormatName,
    options: ConversionOptions,
    settings: Settings,
    cache: CacheManager,
    scratch_dir: Path,
) -> AsyncIterator[bytes]:
    """Run the full conversion pipeline for one request.

    Yields chunks of the merged PDF. Stores a `ConversionResult` on the
    generator's `result` attribute (set after iteration completes) for the
    caller to read into response headers.

    Raises typed ConversionError subclasses; FastAPI handler maps to HTTP.
    """
    started = time.monotonic()
    chunks_rendered = 0
    subdivision_retries = 0
    cache_hits = 0

    source_sha = _file_sha256(input_path)

    # Final-cache lookup
    if options.cache:
        final_cached = cache.get_final(source_sha)
        if final_cached is not None:
            emit_event(
                "cache_hit",
                level="info",
                layer="final",
                source_sha256=source_sha[:16],
            )
            cache_hits += 1
            async for block in _stream_file(final_cached):
                yield block
            _attach_result(
                convert_job,
                ConversionResult(
                    chunks_rendered=0,
                    subdivision_retries=0,
                    cache_hits=1,
                    duration_seconds=time.monotonic() - started,
                ),
            )
            return

    # Probe
    emit_event("probe_start", level="info", format=format)
    probe_started = time.monotonic()
    probe_result = await do_probe(input_path, format, settings, request_id)
    # The probe may correct the format (e.g., file detected as xlsx but is
    # actually a docx — the probe retries and returns the correct format).
    # Use the probe's format for all subsequent dispatch.
    format = probe_result.format
    emit_event(
        "probe_complete",
        level="info",
        page_count=probe_result.page_count,
        natural_seams=len(probe_result.natural_seams),
        size_bytes=probe_result.size_bytes,
        duration_s=round(time.monotonic() - probe_started, 3),
    )

    # Plan chunks. Use adaptive chunk sizing: compute the optimal pages-per-chunk
    # based on file size, page count, format, and RAM budget. The static config
    # value acts as a ceiling (operator override). XLSX gets a per-format floor
    # on max_pages_per_chunk: each Cells subprocess pays a fixed Workbook.Load +
    # pagination cost on the full workbook before rendering its slice, so the
    # chunk size has to be coarse enough to amortize that overhead.
    adaptive_pages = adaptive_max_pages(
        probe_result,
        worker_ram_bytes=settings.worker_ram_bytes,
        parallel=settings.parallel,
    )
    # The static config acts as a ceiling — operator can always cap it lower
    effective_max_pages = min(adaptive_pages, settings.max_pages_per_chunk)
    if probe_result.format == "xlsx":
        effective_max_pages = max(effective_max_pages, settings.xlsx_min_pages_per_chunk)
    elif probe_result.format == "pptx":
        effective_max_pages = max(effective_max_pages, settings.pptx_min_pages_per_chunk)
    emit_event(
        "adaptive_chunk_sizing",
        level="info",
        adaptive_pages=adaptive_pages,
        effective_max_pages=effective_max_pages,
        format=probe_result.format,
    )
    plan = plan_chunks(
        probe_result,
        max_pages_per_chunk=effective_max_pages,
        max_mb_per_chunk=settings.max_mb_per_chunk,
    )
    # Classify the strategy chosen by the planner so logs make it obvious
    # whether intra-request parallelism kicked in.
    if len(plan.chunks) == 1:
        strategy = "single_chunk"
    elif any(c.natural_seam for c in plan.chunks):
        strategy = "natural_seams"
    else:
        strategy = "page_range_split"
    emit_event(
        "plan_complete",
        level="info",
        chunks=len(plan.chunks),
        total_pages=plan.total_pages,
        strategy=strategy,
        parallel=settings.parallel,
    )

    # Dispatch chunk renders under per-job concurrency budget.
    # Two modes: pool (persistent workers, document loaded once) or one-shot
    # (subprocess per chunk). Pool mode eliminates per-chunk document-load
    # overhead but requires C++ worker support for --mode=pool.
    job_sem = asyncio.Semaphore(settings.parallel)
    counters = _Counters()
    use_pool = pool_mode_available(settings, format) and len(plan.chunks) > 1

    if use_pool:
        emit_event("dispatch_mode", level="info", mode="pool", workers=settings.parallel)
        pool_size = min(settings.parallel, len(plan.chunks))
        async with WorkerPool(format, input_path, settings, pool_size=pool_size) as pool:
            # Re-plan if the actual page count differs from the estimate
            # (pool workers report the real count after loading the document)
            if pool.actual_page_count is not None and pool.actual_page_count != plan.total_pages:
                emit_event(
                    "replan_from_pool",
                    level="info",
                    estimated_pages=plan.total_pages,
                    actual_pages=pool.actual_page_count,
                )
                actual_probe = ProbeResult(
                    page_count=pool.actual_page_count,
                    format=probe_result.format,
                    natural_seams=(),
                    size_bytes=probe_result.size_bytes,
                )
                plan = plan_chunks(
                    actual_probe,
                    max_pages_per_chunk=effective_max_pages,
                    max_mb_per_chunk=settings.max_mb_per_chunk,
                )

            async def _render_one_pooled(chunk: Chunk) -> tuple[Chunk, Path]:
                async with job_sem:
                    chunk_key = chunk_sha256(chunk, source_sha, format)
                    cached = cache.get_chunk(chunk_key) if options.cache else None
                    if cached is not None:
                        counters.cache_hits += 1
                        return chunk, cached
                    path = await pool.render_chunk(chunk, scratch_dir)
                    counters.rendered += 1
                    if options.cache:
                        cache.put_chunk(chunk_key, path)
                    return chunk, path

            rendered = await asyncio.gather(
                *(_render_one_pooled(c) for c in plan.chunks)
            )
    else:
        emit_event("dispatch_mode", level="info", mode="one_shot", workers=settings.parallel)

        async def _render_one(chunk: Chunk) -> tuple[Chunk, Path]:
            async with job_sem:
                return chunk, await _render_with_retry(
                    chunk=chunk,
                    input_path=input_path,
                    format=format,
                    scratch_dir=scratch_dir,
                    request_id=request_id,
                    settings=settings,
                    source_sha=source_sha,
                    cache=cache if options.cache else CacheManager(None, settings.aspose_version),
                    counters=counters,
                )

        rendered = await asyncio.gather(
            *(_render_one(c) for c in plan.chunks)
        )
    chunks_rendered = counters.rendered
    subdivision_retries = counters.subdivisions
    cache_hits += counters.cache_hits

    # Order chunk PDFs by chunk index (preserved by gather; here for safety)
    rendered.sort(key=lambda pair: pair[0].index)
    chunk_paths = [p for _, p in rendered]

    # Tee-to-cache: stream qpdf concat into the response AND optionally to cache.
    cache_temp = cache.final_temp_path(source_sha) if options.cache else None

    emit_event("merge_start", level="info", chunk_count=len(chunk_paths))
    merge_started = time.monotonic()
    output_bytes = 0
    try:
        async for block in qpdf.concat_streaming(
            chunk_paths,
            cache_temp_path=cache_temp,
        ):
            output_bytes += len(block)
            yield block
    except BaseException:
        if cache_temp is not None:
            with suppress(OSError):
                cache_temp.unlink(missing_ok=True)
        raise

    emit_event(
        "merge_complete",
        level="info",
        chunk_count=len(chunk_paths),
        output_bytes=output_bytes,
        duration_s=round(time.monotonic() - merge_started, 3),
    )

    if cache_temp is not None:
        cache.finalize_final(source_sha, cache_temp)

    duration = time.monotonic() - started
    emit_event(
        "request_complete",
        level="info",
        chunks_rendered=chunks_rendered,
        subdivision_retries=subdivision_retries,
        cache_hits=cache_hits,
        output_bytes=output_bytes,
        duration_seconds=round(duration, 3),
    )
    _attach_result(
        convert_job,
        ConversionResult(
            chunks_rendered=chunks_rendered,
            subdivision_retries=subdivision_retries,
            cache_hits=cache_hits,
            duration_seconds=duration,
        ),
    )


class _Counters:
    __slots__ = ("rendered", "subdivisions", "cache_hits")

    def __init__(self) -> None:
        self.rendered = 0
        self.subdivisions = 0
        self.cache_hits = 0


async def _render_with_retry(
    *,
    chunk: Chunk,
    input_path: Path,
    format: FormatName,
    scratch_dir: Path,
    request_id: str,
    settings: Settings,
    source_sha: str,
    cache: CacheManager,
    counters: _Counters,
    depth: int = 0,
) -> Path:
    """Render a chunk; on OOM, subdivide and recurse, then concat sub-PDFs.

    Returns a single PDF path for `chunk`. If subdivision was used, the
    returned path is itself a concatenation of sub-PDFs (so the caller
    sees one PDF per planned chunk).
    """
    chunk_key = chunk_sha256(chunk, source_sha, format)
    cached = cache.get_chunk(chunk_key)
    if cached is not None:
        counters.cache_hits += 1
        emit_event(
            "cache_hit",
            level="info",
            layer="chunk",
            chunk_index=chunk.index,
        )
        return cached

    try:
        emit_event(
            "chunk_render_start",
            level="info",
            chunk_index=chunk.index,
            page_range=list(chunk.page_range),
            page_count=chunk.pages,
            depth=depth,
            worker=format,
        )
        chunk_started = time.monotonic()
        path = await aspose_worker.render_chunk(
            chunk=chunk,
            input_path=input_path,
            format=format,
            scratch_dir=scratch_dir,
            request_id=request_id,
            settings=settings,
        )
        chunk_duration = round(time.monotonic() - chunk_started, 3)
        counters.rendered += 1
        try:
            output_bytes = path.stat().st_size
        except OSError:
            output_bytes = -1
        emit_event(
            "chunk_complete",
            level="info",
            chunk_index=chunk.index,
            page_range=list(chunk.page_range),
            depth=depth,
            duration_s=chunk_duration,
            output_bytes=output_bytes,
        )
        cache.put_chunk(chunk_key, path)
        return path
    except OOMError:
        counters.subdivisions += 1
        sub_chunks = subdivide(chunk)
        if not sub_chunks:
            raise SubdivisionFloorError(chunk=chunk, attempts=depth + 1) from None
        emit_event(
            "subdivision_retry",
            level="warn",
            chunk_index=chunk.index,
            page_range_before=list(chunk.page_range),
            sub_count=len(sub_chunks),
            depth=depth,
        )
        sub_paths = await asyncio.gather(
            *(
                _render_with_retry(
                    chunk=sc,
                    input_path=input_path,
                    format=format,
                    scratch_dir=scratch_dir,
                    request_id=request_id,
                    settings=settings,
                    source_sha=source_sha,
                    cache=cache,
                    counters=counters,
                    depth=depth + 1,
                )
                for sc in sub_chunks
            )
        )
        # Merge sub-chunks into one PDF for this chunk's slot.
        merged_path = scratch_dir / f"chunk-{chunk.index}-merged.pdf"
        await _merge_to_file(sub_paths, merged_path)
        return merged_path


async def _merge_to_file(chunk_paths: list[Path], output_path: Path) -> None:
    """Use qpdf to concat chunks into a single file (no streaming needed)."""
    from office_convert import qpdf as qpdf_mod

    with output_path.open("wb") as out:
        async for block in qpdf_mod.concat_streaming(chunk_paths):
            out.write(block)


async def _stream_file(path: Path) -> AsyncIterator[bytes]:
    """Stream a file in 64 KB chunks (for cache-hit responses)."""
    loop = asyncio.get_event_loop()
    with path.open("rb") as f:
        while True:
            block = await loop.run_in_executor(None, f.read, 65536)
            if not block:
                break
            yield block


def _file_sha256(path: Path) -> str:
    h = hashlib.sha256()
    with path.open("rb") as f:
        while True:
            block = f.read(1024 * 1024)
            if not block:
                break
            h.update(block)
    return h.hexdigest()


def _attach_result(fn: object, result: ConversionResult) -> None:
    """Attach ConversionResult to the generator function for caller retrieval.

    We store it as an attribute on the orchestrator module (a module-level
    contextvar would also work). The server reads `convert_job.last_result`
    immediately after the generator exhausts.
    """
    # Simple module-level slot indexed by the running asyncio task.
    task = asyncio.current_task()
    if task is None:
        return
    _RESULTS[id(task)] = result


_RESULTS: dict[int, ConversionResult] = {}


def consume_result(task_id: int) -> ConversionResult | None:
    """Server reads (and clears) the ConversionResult for a completed task."""
    return _RESULTS.pop(task_id, None)

File: internal/orchestrator/orchestrator.go
// Package orchestrator runs the per-request pipeline:
// probe -> plan -> dispatch -> merge -> stream.
//
// Ported from office_convert/orchestrator.py. Implements FR-3..FR-7, FR-9,
// FR-10 — the end-to-end pipeline from business-logic-model.md §1.
//
// The Python original is an async generator yielding PDF bytes and stashes the
// ConversionResult via a per-task module slot. In Go, ConvertJob writes blocks
// to the caller's io.Writer (the http.Flusher-backed response) and simply
// RETURNS the ConversionResult — no contextvar/task-slot trick needed.
package orchestrator

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"io"
	"os"
	"sort"
	"sync"
	"time"

	"github.com/opus2/office-convert-orchestrator/internal/cache"
	"github.com/opus2/office-convert-orchestrator/internal/config"
	"github.com/opus2/office-convert-orchestrator/internal/obs"
	"github.com/opus2/office-convert-orchestrator/internal/oclog"
	"github.com/opus2/office-convert-orchestrator/internal/oerrors"
	"github.com/opus2/office-convert-orchestrator/internal/planner"
	"github.com/opus2/office-convert-orchestrator/internal/probe"
	"github.com/opus2/office-convert-orchestrator/internal/qpdf"
	"github.com/opus2/office-convert-orchestrator/internal/types"
	"github.com/opus2/office-convert-orchestrator/internal/worker"
)

// Deps bundles the shared services a conversion needs.
type Deps struct {
	Settings *config.Settings
	Cache    *cache.Manager
	Stores   worker.Stores
}

// ConvertJob runs the full conversion pipeline for one request, streaming the
// merged PDF into dst (flushing after each block via flush, if non-nil) and
// returning the ConversionResult for the caller's response headers.
//
// Mirrors orchestrator.convert_job.
func ConvertJob(
	ctx context.Context,
	requestID, inputPath string,
	format types.FormatName,
	opts types.ConversionOptions,
	deps Deps,
	scratchDir string,
	dst io.Writer,
	flush func(),
) (*types.ConversionResult, error) {
	s := deps.Settings
	progress := deps.Stores.Progress
	started := time.Now()
	counters := &counters{}

	cacheActive := opts.Cache && deps.Cache.Enabled()
	sourceSHA := ""
	if cacheActive {
		var err error
		sourceSHA, err = fileSHA256(inputPath)
		if err != nil {
			return nil, oerrors.NewInputUnprocessable("hash failed: " + err.Error())
		}
	}

	out := &countingWriter{w: dst, flush: flush}

	// Final-cache lookup.
	if cacheActive {
		if final := deps.Cache.GetFinal(sourceSHA); final != "" {
			oclog.EmitEvent("info", "cache_hit", map[string]any{"layer": "final", "source_sha256": truncate(sourceSHA, 16)})
			progress.Update(requestID, obs.ProgressUpdate{
				TotalChunks: intp(1), LoadProgress: f64p(1.0), MergeDone: f64p(1.0), Phase: strp("complete"),
			})
			if err := streamFile(final, out); err != nil {
				return nil, err
			}
			return &types.ConversionResult{CacheHits: 1, DurationSeconds: time.Since(started).Seconds()}, nil
		}
	}

	// Probe.
	oclog.EmitEvent("info", "probe_start", map[string]any{"format": format})
	probeStarted := time.Now()
	pr, err := probe.Probe(ctx, s, inputPath, format, requestID)
	if err != nil {
		return nil, err
	}
	format = pr.Format // probe may correct the format
	oclog.EmitEvent("info", "probe_complete", map[string]any{
		"page_count": pr.PageCount, "natural_seams": len(pr.NaturalSeams),
		"size_bytes": pr.SizeBytes, "duration_s": round3(time.Since(probeStarted)),
	})

	// Plan with adaptive chunk sizing.
	adaptivePages := planner.AdaptiveMaxPages(pr, s.WorkerRAMBytes, s.Parallel, planner.AdaptiveMaxPagesOptions{})
	effMaxPages := minInt(adaptivePages, s.MaxPagesPerChunk)
	switch format {
	case types.FormatXLSX:
		effMaxPages = maxInt(effMaxPages, s.XLSXMinPagesPerChunk)
	case types.FormatPPTX:
		effMaxPages = maxInt(effMaxPages, s.PPTXMinPagesPerChunk)
	}
	oclog.EmitEvent("info", "adaptive_chunk_sizing", map[string]any{
		"adaptive_pages": adaptivePages, "effective_max_pages": effMaxPages, "format": format,
	})
	plan := planner.PlanChunks(pr, effMaxPages, s.MaxMBPerChunk)
	oclog.EmitEvent("info", "plan_complete", map[string]any{
		"chunks": len(plan.Chunks), "total_pages": plan.TotalPages,
		"strategy": planStrategy(plan), "parallel": s.Parallel,
	})
	progress.Update(requestID, obs.ProgressUpdate{TotalChunks: intp(len(plan.Chunks)), Phase: strp("loading")})

	usePool := worker.PoolModeAvailable() && len(plan.Chunks) >= s.PoolMinChunks

	var rendered []chunkPath
	if usePool {
		rendered, plan, err = dispatchPool(ctx, requestID, inputPath, format, deps, scratchDir, plan, effMaxPages, pr, sourceSHA, cacheActive, counters)
	} else {
		rendered, err = dispatchOneShot(ctx, requestID, inputPath, format, deps, scratchDir, plan, sourceSHA, cacheActive, counters)
	}
	if err != nil {
		return nil, err
	}

	// Order chunk PDFs by chunk index.
	sort.Slice(rendered, func(i, j int) bool { return rendered[i].chunk.Index < rendered[j].chunk.Index })
	chunkPaths := make([]string, len(rendered))
	for i, r := range rendered {
		chunkPaths[i] = r.path
	}

	// Stream qpdf concat into the response (and optionally tee to cache).
	cacheTemp := ""
	if cacheActive {
		cacheTemp, _ = deps.Cache.FinalTempPath(sourceSHA)
	}
	oclog.EmitEvent("info", "merge_start", map[string]any{"chunk_count": len(chunkPaths)})
	progress.Update(requestID, obs.ProgressUpdate{Phase: strp("merging")})
	mergeStarted := time.Now()
	if err := qpdf.ConcatStreaming(ctx, chunkPaths, out, cacheTemp); err != nil {
		if cacheTemp != "" {
			_ = os.Remove(cacheTemp)
		}
		return nil, err
	}
	oclog.EmitEvent("info", "merge_complete", map[string]any{
		"chunk_count": len(chunkPaths), "output_bytes": out.n, "duration_s": round3(time.Since(mergeStarted)),
	})
	progress.Update(requestID, obs.ProgressUpdate{MergeDone: f64p(1.0), Phase: strp("complete")})

	if cacheTemp != "" {
		_ = deps.Cache.FinalizeFinal(sourceSHA, cacheTemp)
	}

	dur := time.Since(started).Seconds()
	r, sub, ch := counters.snapshot()
	oclog.EmitEvent("info", "request_complete", map[string]any{
		"chunks_rendered": r, "subdivision_retries": sub, "cache_hits": ch,
		"output_bytes": out.n, "duration_seconds": round3time(dur),
	})
	return &types.ConversionResult{
		ChunksRendered: r, SubdivisionRetries: sub, CacheHits: ch, DurationSeconds: dur,
	}, nil
}

type chunkPath struct {
	chunk types.Chunk
	path  string
}

// dispatchPool renders all chunks via a persistent pool (forked or legacy).
// Returns the rendered chunk paths and the (possibly re-planned) plan.
func dispatchPool(
	ctx context.Context, requestID, inputPath string, format types.FormatName, deps Deps,
	scratchDir string, plan types.ChunkPlan, effMaxPages int, pr types.ProbeResult,
	sourceSHA string, cacheActive bool, counters *counters,
) ([]chunkPath, types.ChunkPlan, error) {
	s := deps.Settings
	progress := deps.Stores.Progress

	poolSize := minInt(s.Parallel, len(plan.Chunks))
	forked := worker.ForkAfterLoadEnabled(s, format)
	if format == types.FormatXLSX && !forked {
		poolSize = minInt(poolSize, s.XLSXMaxPoolSize)
	}

	var pool worker.Pool
	var err error
	if forked {
		oclog.EmitEvent("info", "dispatch_mode", map[string]any{"mode": "pool_fork", "workers": s.Parallel, "pool_size": poolSize})
		pool, err = worker.NewForkedWorkerPool(s, format, inputPath, requestID, deps.Stores, poolSize)
	} else {
		oclog.EmitEvent("info", "dispatch_mode", map[string]any{"mode": "pool", "workers": s.Parallel})
		pool, err = worker.NewWorkerPool(s, format, inputPath, requestID, deps.Stores, poolSize)
	}
	if err != nil {
		return nil, plan, err
	}
	defer pool.Close(context.Background())

	progress.Update(requestID, obs.ProgressUpdate{LoadProgress: f64p(1.0), Phase: strp("rendering")})

	// Re-plan if the actual page count differs from the estimate.
	if actual, ok := pool.ActualPageCount(); ok && actual != plan.TotalPages {
		oclog.EmitEvent("info", "replan_from_pool", map[string]any{"estimated_pages": plan.TotalPages, "actual_pages": actual})
		actualProbe := types.ProbeResult{PageCount: actual, Format: pr.Format, SizeBytes: pr.SizeBytes}
		plan = planner.PlanChunks(actualProbe, effMaxPages, s.MaxMBPerChunk)
		progress.Update(requestID, obs.ProgressUpdate{TotalChunks: intp(len(plan.Chunks))})
	}

	render := func(ctx context.Context, chunk types.Chunk) (string, error) {
		if cacheActive {
			key := planner.ChunkSHA256(chunk, sourceSHA, format)
			if cached := deps.Cache.GetChunk(key); cached != "" {
				counters.addCacheHit()
				progress.Update(requestID, obs.ProgressUpdate{IncrementChunks: 1})
				return cached, nil
			}
			path, err := pool.RenderChunk(ctx, chunk, scratchDir)
			if err != nil {
				return "", err
			}
			counters.addRendered()
			_ = deps.Cache.PutChunk(key, path)
			progress.Update(requestID, obs.ProgressUpdate{IncrementChunks: 1})
			return path, nil
		}
		path, err := pool.RenderChunk(ctx, chunk, scratchDir)
		if err != nil {
			return "", err
		}
		counters.addRendered()
		progress.Update(requestID, obs.ProgressUpdate{IncrementChunks: 1})
		return path, nil
	}

	rendered, err := renderAll(ctx, plan.Chunks, s.Parallel, render)
	return rendered, plan, err
}

// dispatchOneShot renders all chunks via per-chunk subprocesses with OOM
// subdivision retry.
func dispatchOneShot(
	ctx context.Context, requestID, inputPath string, format types.FormatName, deps Deps,
	scratchDir string, plan types.ChunkPlan, sourceSHA string, cacheActive bool, counters *counters,
) ([]chunkPath, error) {
	s := deps.Settings
	oclog.EmitEvent("info", "dispatch_mode", map[string]any{"mode": "one_shot", "workers": s.Parallel})
	deps.Stores.Progress.Update(requestID, obs.ProgressUpdate{LoadProgress: f64p(1.0), Phase: strp("rendering")})

	render := func(ctx context.Context, chunk types.Chunk) (string, error) {
		return renderWithRetry(ctx, chunk, inputPath, format, deps, scratchDir, requestID, sourceSHA, cacheActive, counters, 0)
	}
	return renderAll(ctx, plan.Chunks, s.Parallel, render)
}

// renderWithRetry renders a chunk; on OOM it subdivides and recurses, then
// concatenates the sub-PDFs into one. Mirrors orchestrator._render_with_retry.
func renderWithRetry(
	ctx context.Context, chunk types.Chunk, inputPath string, format types.FormatName, deps Deps,
	scratchDir, requestID, sourceSHA string, cacheActive bool, counters *counters, depth int,
) (string, error) {
	s := deps.Settings
	if cacheActive {
		key := planner.ChunkSHA256(chunk, sourceSHA, format)
		if cached := deps.Cache.GetChunk(key); cached != "" {
			counters.addCacheHit()
			oclog.EmitEvent("info", "cache_hit", map[string]any{"layer": "chunk", "chunk_index": chunk.Index})
			return cached, nil
		}
	}

	oclog.EmitEvent("info", "chunk_render_start", map[string]any{
		"chunk_index": chunk.Index, "page_range": []int{chunk.PageStart, chunk.PageEnd},
		"page_count": chunk.Pages(), "depth": depth, "worker": format,
	})
	chunkStarted := time.Now()
	path, err := worker.RenderChunkOneShot(ctx, s, chunk, inputPath, format, scratchDir, requestID)
	if err != nil {
		if oe, ok := err.(*oerrors.Error); ok && oe.OOM {
			counters.addSubdivision()
			subs := planner.Subdivide(chunk)
			if len(subs) == 0 {
				return "", oerrors.NewSubdivisionFloor(chunk, depth+1)
			}
			oclog.EmitEvent("warn", "subdivision_retry", map[string]any{
				"chunk_index": chunk.Index, "page_range_before": []int{chunk.PageStart, chunk.PageEnd},
				"sub_count": len(subs), "depth": depth,
			})
			subPaths, serr := renderAll(ctx, subs, len(subs), func(ctx context.Context, sc types.Chunk) (string, error) {
				return renderWithRetry(ctx, sc, inputPath, format, deps, scratchDir, requestID, sourceSHA, cacheActive, counters, depth+1)
			})
			if serr != nil {
				return "", serr
			}
			merged := fmt.Sprintf("%s/chunk-%s-merged.pdf", scratchDir, formatIndex(chunk.Index))
			paths := make([]string, len(subPaths))
			for i, sp := range subPaths {
				paths[i] = sp.path
			}
			if err := qpdf.ConcatToFile(ctx, paths, merged); err != nil {
				return "", err
			}
			return merged, nil
		}
		return "", err
	}

	counters.addRendered()
	deps.Stores.Progress.Update(requestID, obs.ProgressUpdate{IncrementChunks: 1})
	outBytes := int64(-1)
	if info, err := os.Stat(path); err == nil {
		outBytes = info.Size()
	}
	oclog.EmitEvent("info", "chunk_complete", map[string]any{
		"chunk_index": chunk.Index, "page_range": []int{chunk.PageStart, chunk.PageEnd},
		"depth": depth, "duration_s": round3(time.Since(chunkStarted)), "output_bytes": outBytes,
	})
	if cacheActive {
		_ = deps.Cache.PutChunk(planner.ChunkSHA256(chunk, sourceSHA, format), path)
	}
	return path, nil
}

// renderAll runs renderFn over chunks with a bounded concurrency of `parallel`,
// preserving input order in the result and failing fast on the first error.
// Mirrors the asyncio.Semaphore(parallel) + gather pattern.
func renderAll(ctx context.Context, chunks []types.Chunk, parallel int, renderFn func(context.Context, types.Chunk) (string, error)) ([]chunkPath, error) {
	if parallel < 1 {
		parallel = 1
	}
	ctx, cancel := context.WithCancel(ctx)
	defer cancel()

	out := make([]chunkPath, len(chunks))
	sem := make(chan struct{}, parallel)
	var wg sync.WaitGroup
	var mu sync.Mutex
	var firstErr error

	for i, c := range chunks {
		wg.Add(1)
		go func(i int, c types.Chunk) {
			defer wg.Done()
			select {
			case sem <- struct{}{}:
			case <-ctx.Done():
				return
			}
			defer func() { <-sem }()
			if ctx.Err() != nil {
				return
			}
			p, err := renderFn(ctx, c)
			if err != nil {
				mu.Lock()
				if firstErr == nil {
					firstErr = err
					cancel()
				}
				mu.Unlock()
				return
			}
			out[i] = chunkPath{chunk: c, path: p}
		}(i, c)
	}
	wg.Wait()
	if firstErr != nil {
		return nil, firstErr
	}
	return out, nil
}

// --- counters (concurrent-safe; Python relied on the GIL) ---

type counters struct {
	mu                       sync.Mutex
	rendered, subs, cacheHit int
}

func (c *counters) addRendered()    { c.mu.Lock(); c.rendered++; c.mu.Unlock() }
func (c *counters) addSubdivision() { c.mu.Lock(); c.subs++; c.mu.Unlock() }
func (c *counters) addCacheHit()    { c.mu.Lock(); c.cacheHit++; c.mu.Unlock() }
func (c *counters) snapshot() (int, int, int) {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.rendered, c.subs, c.cacheHit
}

// --- helpers ---

func planStrategy(plan types.ChunkPlan) string {
	if len(plan.Chunks) == 1 {
		return "single_chunk"
	}
	for _, c := range plan.Chunks {
		if c.NaturalSeam {
			return "natural_seams"
		}
	}
	return "page_range_split"
}

// countingWriter counts bytes and flushes after each write.
type countingWriter struct {
	w     io.Writer
	flush func()
	n     int64
}

func (c *countingWriter) Write(p []byte) (int, error) {
	n, err := c.w.Write(p)
	c.n += int64(n)
	if c.flush != nil {
		c.flush()
	}
	return n, err
}

func streamFile(path string, dst io.Writer) error {
	f, err := os.Open(path)
	if err != nil {
		return err
	}
	defer f.Close()
	buf := make([]byte, 65536)
	_, err = io.CopyBuffer(dst, f, buf)
	return err
}

func fileSHA256(path string) (string, error) {
	f, err := os.Open(path)
	if err != nil {
		return "", err
	}
	defer f.Close()
	h := sha256.New()
	if _, err := io.Copy(h, f); err != nil {
		return "", err
	}
	return hex.EncodeToString(h.Sum(nil)), nil
}

func formatIndex(idx float64) string {
	s := fmt.Sprintf("%g", idx)
	for _, r := range s {
		if r == '.' || r == 'e' || r == 'E' {
			return s
		}
	}
	return s + ".0"
}

func round3(d time.Duration) float64 { return round3time(d.Seconds()) }
func round3time(sec float64) float64 { return float64(int64(sec*1000+0.5)) / 1000 }
func truncate(s string, n int) string {
	if len(s) > n {
		return s[:n]
	}
	return s
}
func minInt(a, b int) int {
	if a < b {
		return a
	}
	return b
}
func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}
func intp(i int) *int         { return &i }
func f64p(f float64) *float64 { return &f }
func strp(s string) *string   { return &s }
